File: backend/app/schemas/security.py
from datetime import datetime
from ipaddress import IPv4Address, IPv6Address
from typing import Optional
from uuid import UUID

from pydantic import BaseModel

from app.schemas.common import BaseResponseSchema

from app.models.security.audit_log import (
    AuditAction,
    EntityType,
)

from app.models.security.blocked_ip import (
    BlockReason,
)

from app.models.security.login_attempt import (
    AttemptType,
    AuthenticationFailureReason,
)

from app.models.security.request_log import (
    HttpMethod,
    HttpProtocol,
    RequestSource,
    RequestOutcome,
)

from app.models.security.security_event import (
    EventSeverity,
    SecurityEventType,
)


# -------------------------------------------------------
# Security Events
# -------------------------------------------------------

class SecurityEventResponse(BaseResponseSchema):
    event_type: SecurityEventType
    severity: EventSeverity
    title: str
    description: str
    ip_address: IPv4Address | IPv6Address
    country: Optional[str] = None
    city: Optional[str] = None
    resolved: bool
    evidence: Optional[dict] = None


# -------------------------------------------------------
# Request Logs
# -------------------------------------------------------

class RequestLogResponse(BaseResponseSchema):
    method: HttpMethod
    protocol: HttpProtocol
    source: RequestSource
    outcome: RequestOutcome
    route: str
    path: str
    query_string: str | None = None
    status_code: int
    response_time_ms: int
    ip_address: IPv4Address | IPv6Address
    user_agent: str
    country: str | None = None
    city: str | None = None
    customer_id: UUID | None = None


class RequestLogListResponse(BaseModel):
    items: list[RequestLogResponse]
    total: int
    limit: int
    offset: int


# -------------------------------------------------------
# Customer Sessions
# -------------------------------------------------------

class CustomerSessionResponse(BaseResponseSchema):
    customer_id: UUID

    device_name: str | None = None
    browser: str | None = None
    operating_system: str | None = None

    ip_address: IPv4Address | IPv6Address
    user_agent: str

    country: str | None = None
    city: str | None = None

    login_at: datetime
    last_activity: datetime
    expires_at: datetime
    revoked_at: datetime | None = None


class CustomerSessionListResponse(BaseModel):
    items: list[CustomerSessionResponse]
    total: int
    limit: int
    offset: int


# -------------------------------------------------------
# Login Attempts
# -------------------------------------------------------

class LoginAttemptResponse(BaseResponseSchema):
    email: str
    ip_address: IPv4Address | IPv6Address
    user_agent: str
    successful: bool
    failure_reason: AuthenticationFailureReason | None = None
    attempt_type: AttemptType
    customer_id: UUID | None = None


class LoginAttemptListResponse(BaseModel):
    items: list[LoginAttemptResponse]
    total: int
    limit: int
    offset: int


# -------------------------------------------------------
# Audit Logs
# -------------------------------------------------------

class AuditLogResponse(BaseResponseSchema):
    action: AuditAction
    entity_type: EntityType
    entity_name: str | None = None
    entity_public_id: str | None = None
    old_data: dict | None = None
    new_data: dict | None = None
    ip_address: IPv4Address | IPv6Address
    user_agent: str
    customer_id: UUID | None = None


class AuditLogListResponse(BaseModel):
    items: list[AuditLogResponse]
    total: int
    limit: int
    offset: int


# -------------------------------------------------------
# Blocked IPs
# -------------------------------------------------------

class BlockedIPResponse(BaseResponseSchema):
    ip_address: IPv4Address | IPv6Address
    reason: BlockReason
    blocked_by: str
    block_note: Optional[str] = None
    blocked_until: Optional[datetime] = None
    permanently_blocked: bool
    is_active: bool


class BlockedIPListResponse(BaseModel):
    items: list[BlockedIPResponse]
    total: int
    limit: int
    offset: int


class BlockedIPCreateSchema(BaseModel):
    ip_address: IPv4Address | IPv6Address
    reason: BlockReason
    note: Optional[str] = None
    expires_in_minutes: Optional[int] = None
    permanently: bool = False

File: frontend/src/services/security.ts
import api from "@/lib/axios";

export interface CustomerSession {
  public_id: string;
  customer_id: string;

  device_name: string | null;
  browser: string | null;
  operating_system: string | null;

  ip_address: string;
  user_agent: string;

  country: string | null;
  city: string | null;

  login_at: string;
  last_activity: string;
  expires_at: string;
  revoked_at: string | null;

  created_at: string;
  updated_at: string;
}

export interface CustomerSessionListResponse {
  items: CustomerSession[];
  total: number;
  limit: number;
  offset: number;
}

export interface SecurityEvent {
  public_id: string;
  event_type: string;
  severity: "low" | "medium" | "high" | "critical";
  title: string;
  description: string;
  ip_address: string;
  country: string | null;
  city: string | null;
  resolved: boolean;
  created_at: string;
  evidence: any;
}

export interface RequestLog {
  public_id: string;
  method: string;
  route: string;
  path: string;
  status_code: number;
  response_time_ms: number;
  ip_address: string;
  user_agent: string;
  created_at: string;
  customer_id: string | null;
  query_string: string | null;
}

export interface LoginAttempt {
  public_id: string;
  email: string;
  ip_address: string;
  user_agent: string;
  successful: boolean;
  failure_reason: string | null;
  attempt_type: string;
  created_at: string;
  customer_id: string | null;
}

export interface AuditLog {
  public_id: string;
  action: string;
  entity_type: string;
  entity_name: string | null;
  entity_public_id: string | null;
  old_data: any;
  new_data: any;
  ip_address: string;
  user_agent: string;
  created_at: string;
  customer_id: string | null;
}

export interface BlockedIP {
  public_id: string;
  ip_address: string;
  reason: string;
  blocked_by: string;
  block_note: string | null;
  blocked_until: string | null;
  permanently_blocked: boolean;
  is_active: boolean;
  created_at: string;
}

export interface DashboardStats {
  total_requests: number;
  total_security_events: number;
  critical_events: number;
  high_events: number;
  active_blocked_ips: number;
  failed_logins_last_24h: number;
}

export const securityService = {
  // ---------------------------------------------------------
  // Dashboard
  // ---------------------------------------------------------

  async getDashboardStats(): Promise<DashboardStats> {
    const response = await api.get(
      "/admin/security/dashboard/stats"
    );

    return response.data;
  },

  // ---------------------------------------------------------
  // Security Events
  // ---------------------------------------------------------

  async listEvents(params?: {
    severity?: string;
    resolved?: boolean;
    limit?: number;
  }): Promise<SecurityEvent[]> {
    const response = await api.get(
      "/admin/security/events",
      {
        params,
      }
    );

    return response.data;
  },

  async resolveEvent(
    publicId: string
  ): Promise<void> {
    await api.post(
      `/admin/security/events/${publicId}/resolve`
    );
  },

  // ---------------------------------------------------------
  // Customer Sessions
  // ---------------------------------------------------------

  async getCustomerSessions(
    params: {
      limit?: number;
      offset?: number;
    } = {}
  ): Promise<CustomerSessionListResponse> {
    const response = await api.get(
      "/admin/security/sessions",
      {
        params,
      }
    );

    return response.data;
  },

  async revokeCustomerSession(
    publicId: string
  ): Promise<CustomerSession> {
    const response = await api.post(
      `/admin/security/sessions/${publicId}/revoke`
    );

    return response.data;
  },

  // ---------------------------------------------------------
  // Request Logs
  // ---------------------------------------------------------

  async getRequestLogs(
    params: {
      method?: string;
      status_code?: number;
      ip?: string;
      start_date?: string;
      end_date?: string;
      limit?: number;
      offset?: number;
    } = {}
  ): Promise<{
    items: RequestLog[];
    total: number;
  }> {
    const response = await api.get(
      "/admin/security/requests",
      {
        params,
      }
    );

    return response.data;
  },

  // ---------------------------------------------------------
  // Login Attempts
  // ---------------------------------------------------------

  async getLoginAttempts(
    params: {
      email?: string;
      ip?: string;
      successful?: boolean;
      start_date?: string;
      end_date?: string;
      limit?: number;
      offset?: number;
    } = {}
  ): Promise<{
    items: LoginAttempt[];
    total: number;
  }> {
    const response = await api.get(
      "/admin/security/login-attempts",
      {
        params,
      }
    );

    return response.data;
  },

  // ---------------------------------------------------------
  // Audit Logs
  // ---------------------------------------------------------

  async getAuditLogs(
    params: {
      action?: string;
      entity_type?: string;
      customer_id?: string;
      start_date?: string;
      end_date?: string;
      limit?: number;
      offset?: number;
    } = {}
  ): Promise<{
    items: AuditLog[];
    total: number;
  }> {
    const response = await api.get(
      "/admin/security/audit-logs",
      {
        params,
      }
    );

    return response.data;
  },

  // ---------------------------------------------------------
  // Blocked IPs
  // ---------------------------------------------------------

  async getBlockedIPs(
    params: {
      active_only?: boolean;
      limit?: number;
      offset?: number;
    } = {}
  ): Promise<{
    items: BlockedIP[];
    total: number;
  }> {
    const response = await api.get(
      "/admin/security/blocked-ips",
      {
        params,
      }
    );

    return response.data;
  },

  async blockIp(data: {
    ip_address: string;
    reason: string;
    note?: string;
    expires_in_minutes?: number;
    permanently?: boolean;
  }): Promise<{
    public_id: string;
  }> {
    const response = await api.post(
      "/admin/security/blocked-ips",
      data
    );

    return response.data;
  },

  async unblockIp(
    publicId: string
  ): Promise<void> {
    await api.delete(
      `/admin/security/blocked-ips/${publicId}`
    );
  },

  // ---------------------------------------------------------
  // Statistics
  // ---------------------------------------------------------

  async getTopIPsByRequests(
    since?: string
  ): Promise<
    {
      ip: string;
      count: number;
    }[]
  > {
    const response = await api.get(
      "/admin/security/stats/requests-by-ip",
      {
        params: {
          since,
        },
      }
    );

    return response.data;
  },

  async getTopIPsByFailedLogins(
    since?: string
  ): Promise<
    {
      ip: string;
      count: number;
    }[]
  > {
    const response = await api.get(
      "/admin/security/stats/failed-logins-by-ip",
      {
        params: {
          since,
        },
      }
    );

    return response.data;
  },
};